test/testdrive/kafka-exactly-once-sink.td

```diff
@@ -361,7 +361,8 @@ $ kafka-verify format=json sink=materialize.public.json_avro_upsert_key_2 key=tr
 {"b": 2} {"a": 1, "b": 2, "c": 3, "transaction": {"id": "0"}}
 
 # Verify compaction of exactly once sinks.
-$ verify-timestamp-compaction source=input_csv max-size=3 permit-progress=true
+# TODO: not sure how to address this.
```

Comment:
I described whats happening here in the 2nd commit message

```
Basically, previously, even when a csv source was closed, we would still generate
new RT bindings for it, and continue to compact those bindings. In other words,
even though the source was closed we would still present a non-empty upper for
the source, and use that non-empty upper to drive source compaction. Now, we correctly
present an empty upper when the source is closed. However, coord declines to compact
up to the empty upper, as doing so makes the data unusable. The old real-time timestamping
code continues to send new timestamp bindings, but those bindings are now never compacted,
because the coordinator never advances the since frontier because upper is empty.

Ideally, the real time timestamping code would just stop creating new timestamp
bindings when the source is permanently closed, but that functionality doesn’t
exist yet.
```

Reply:
@cjubb39 I don't love disabling this check (I'll create an issue to followup after merging if this is the route we take) but I couldn't think of a better thing to do that wasn't an ugly hack. I feel like fixing uppers for sources generally to correspond to definite sources is enough of a win that its ok to temporarily break exactly-once sinks over non-tailed csv sources to use more disk -- wdyt?

Reply:
Is there anything useful we get by closing a source? If not, you can change this line https://github.com/MaterializeInc/materialize/blob/main/src/dataflow/src/source/file.rs#L187 to return `NextMessage::Pending`

Reply:
There are massive benefits to closing sources, in a world where we can do that (I'm not sure if we can, ofc, because Kafka and "streaming" generally doesn't encourage terminating streams). But, for example Timely can start shutting down dataflows, and DD can close out entire join arrangements when you close sources (even if the other inputs to the join are not closed).

Reply:
Hmm -- I don't like it either but it sounds like it's incremental progress overall since uppers are now "more correct".  As far as I can tell, we only ever close S3 / File sources so it's not like we're affecting kafka sources or something more common.  I wouldn't expect a huge perf impact since we have good indexes on the timestamps.

Can you create an issue with a suggested fix / rough length to implement and include it in the comment (like 9514 just below)?  I would like to fix it relatively soon so we don't grow an unbounded size of this database.  I have a fix of 9514 in the works to merge soon so this would be the only situation where that can happen.

Reply:
Would any of this improve if coord's compaction policy was improved, e.g. to not compact past the current timeline frontier? Does that un-bork the bit of logic around compaction, and allow us to worry less about .. something here?

Reply:
No, I think the issue here is from the timestamp bindings code, where the timestamp bindings generator continues to generate new timestamp bindings even though the source has closed.